packages/ts/src/rules/explicitModuleBoundaryTypes.ts

```diff
@@ -0,0 +1,171 @@
+// eslint-disable perfectionist/sort-switch-case -- cases are ordered by importance, not alphabetically
+import type { ReportMessageData } from "@flint.fyi/core";
+import type { RuleContext } from "@flint.fyi/core/src/types/context.js";
+
+import * as ts from "typescript";
+
+import { getTSNodeRange } from "../getTSNodeRange.js";
+import { typescriptLanguage } from "../language.js";
+
+const messages = {
+	fnParam: {
+		primary:
+			"Public facing functions should have explicit types for all parameters.",
+		secondary: ["{{ parameterName }} is declared here."],
+		suggestions: [],
+	},
+	fnReturn: {
+		primary: "Public facing functions should have an explicit return type.",
+		secondary: [],
+		suggestions: [],
+	},
+	fnReturnWithSuggestion: {
+		primary: "Public facing functions should have an explicit return type.",
+		secondary: [],
+		suggestions: [
+			"Add `{{ returnType }}` to the return type of this function.",
+		],
+	},
+} satisfies Record<string, ReportMessageData>;
+
+export default typescriptLanguage.createRule({
+	about: {
+		id: "explicitModuleBoundaryTypes",
+		preset: "restrictive",
+	},
+	messages,
+	setup(ctx) {
+		const sourceFile = ctx.sourceFile;
+		// ignore .d.ts and .js files
+		if (sourceFile.isDeclarationFile || isJavaScriptFile(sourceFile)) {
+			return;
+		}
+		const src = ctx.typeChecker.getSymbolAtLocation(sourceFile);
+		const exports = src?.exports;
+		if (!exports) {
+			return;
+		}
+		checkExports(ctx, exports);
```

Comment:
I think the more "idiomatic" way to do this in the current selector-based rules would be to return a set of selectors:

```ts
return {
  FunctionDeclaration(node) {
    if (tsutils.isModifierSet(node, ts.ModifierFlags.Export) {
      // ...
    }
  },
  ModuleDeclaration(node) {
    if ( /* ... */ ) {
      // ...
    }
  },
  // ...
};

`checkExports` has a `switch` over declaration types anyway, so I think this would be roughly the same amount of code and complexity? But is there a bunch of logic done for us by going through the source file's symbol's exports?

Reply:
I would add an `ExportDeclaration` query and then use a scope manager to jump to definition from `foo` -> `function foo() {}`. But since Flint doesn't have any scope manager yet I'd use the equivalent TypeScript types APIs.